pkg/kubelet/server/stats/fs_resource_analyzer.go

```diff
@@ -0,0 +1,146 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package stats
+
+import (
+	"fmt"
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/types"
+	"k8s.io/kubernetes/pkg/volume"
+	"sync"
+	"time"
+)
+
+// PodVolumeStats encapsulates all VolumeStats for a pod
+type PodVolumeStats struct {
+	PodUid  types.UID
+	Volumes []VolumeStats
+}
+
+// fsResourceAnalyzerInterface is for embedding fs functions into ResourceAnalyzer
+type fsResourceAnalyzerInterface interface {
+	GetPodVolumeStats(uid types.UID) (PodVolumeStats, error)
+}
+
+// diskResourceAnalyzer provider stats about fs resource usage
+type fsResourceAnalyzer struct {
+	volumeCache         sync.RWMutex
+	statsProvider       StatsProvider
+	cachedVolumeStats   *cachedPodVolumeStats // Eagerly cache volume stats
+	calcVolumeFrequency time.Duration         // How frequently to calculate disk usage
+}
+
+var _ fsResourceAnalyzerInterface = &fsResourceAnalyzer{}
+
+// cachedPodVolumeStats caches the results of eagerly calculating PodVolumeStats
+type cachedPodVolumeStats struct {
+	volumeStats map[types.UID]*PodVolumeStats
+}
+
+// default calculation interval to use if none is provided
+const defaultVolumeStatsCalcInterval = 30 * time.Second
+
+// NewStats returns a new NewDiskResourceAnalyzer implementation
+func newFsResourceAnalyzer(statsProvider StatsProvider, calcVolumeFrequency time.Duration) *fsResourceAnalyzer {
+	return &fsResourceAnalyzer{
+		statsProvider:       statsProvider,
+		calcVolumeFrequency: calcVolumeFrequency,
+	}
+}
+
+// Start eager background caching of volume stats.
+func (s *fsResourceAnalyzer) Start() {
+	go func() {
+		sleepTime := s.calcVolumeFrequency
+		if sleepTime <= 0 {
+			glog.Warningf("volume stats calc interval must be greater than 0, defaulting to %v.", defaultVolumeStatsCalcInterval)
+			sleepTime = defaultVolumeStatsCalcInterval
+		}
+		for {
+			glog.V(3).Infof("Calculating volume stats and caching result.")
+			startTime := time.Now()
+			s.updateCachedPodVolumeStats()
+			elapsedDuration := time.Now().Sub(startTime)
+			glog.V(3).Infof("Finished calculating volume stats in %v, sleeping for %v.", elapsedDuration, sleepTime)
+			time.Sleep(sleepTime)
```

Comment:
I would recommend handling this function differently:
1) Use util.Forever instead of manually looping / sleeping
2) Log the latency to prometheus instead of the kubelet log (Or at least make it log level 4)